File: backend/routers/appointments.py
"""
routers/appointments.py — Appointment CRUD endpoints with role-based access control.
"""

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from database import get_db
from models import Appointment, User
from schemas import AppointmentCreate, AppointmentUpdate, AppointmentOut
from dependencies import get_current_user, require_role

router = APIRouter(prefix="/appointments", tags=["Appointments"])


@router.get("", response_model=list[AppointmentOut])
def get_appointments(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Get appointments.
    - Admin / Receptionist: see all appointments.
    - Doctor: see only their own assigned appointments.
    """
    query = db.query(Appointment)

    if current_user.role == "doctor":
        query = query.filter(Appointment.doctor_id == current_user.doctor_id)

    return query.order_by(Appointment.id).all()


@router.post("", response_model=AppointmentOut, status_code=status.HTTP_201_CREATED)
def create_appointment(
    data: AppointmentCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_role("admin", "receptionist")),
):
    """
    Book a new appointment.
    Accessible by: Admin, Receptionist.
    """
    # Slot conflict check: same doctor + date + time (exclude cancelled)
    conflict = db.query(Appointment).filter(
        Appointment.doctor_id == data.doctor_id,
        Appointment.date == data.date,
        Appointment.time == data.time,
        Appointment.status != "Cancelled",
    ).first()
    if conflict:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"This slot is already booked for {conflict.patient_name}",
        )

    appointment = Appointment(
        patient_id=data.patient_id,
        patient_name=data.patient_name,
        doctor_id=data.doctor_id,
        doctor_name=data.doctor_name,
        date=data.date,
        time=data.time,
        type=data.type or "Consultation",
        status=data.status or "Pending",
        notes=data.notes or "",
    )
    db.add(appointment)
    db.commit()
    db.refresh(appointment)
    return appointment


@router.put("/{appointment_id}", response_model=AppointmentOut)
def update_appointment(
    appointment_id: int,
    data: AppointmentUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_role("admin")),
):
    """
    Update an existing appointment.
    Accessible by: Admin only.
    """
    appointment = db.query(Appointment).filter(Appointment.id == appointment_id).first()
    if not appointment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Appointment not found")

    update_data = data.model_dump(exclude_unset=True)

    # Slot conflict check when date/time/doctor changes
    new_doctor = update_data.get("doctor_id", appointment.doctor_id)
    new_date = update_data.get("date", appointment.date)
    new_time = update_data.get("time", appointment.time)
    new_status = update_data.get("status", appointment.status)

    if new_status != "Cancelled":
        conflict = db.query(Appointment).filter(
            Appointment.id != appointment_id,
            Appointment.doctor_id == new_doctor,
            Appointment.date == new_date,
            Appointment.time == new_time,
            Appointment.status != "Cancelled",
        ).first()
        if conflict:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=f"This slot is already booked for {conflict.patient_name}",
            )

    for key, value in update_data.items():
        setattr(appointment, key, value)

    db.commit()
    db.refresh(appointment)
    return appointment


@router.delete("/{appointment_id}", response_model=AppointmentOut)
def cancel_appointment(
    appointment_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_role("admin")),
):
    """
    Cancel an appointment (sets status to 'Cancelled').
    Accessible by: Admin only.
    """
    appointment = db.query(Appointment).filter(Appointment.id == appointment_id).first()
    if not appointment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Appointment not found")

    appointment.status = "Cancelled"
    db.commit()
    db.refresh(appointment)
    return appointment

File: backend/routes/appointments.js
const express = require('express');
const router = express.Router();
const Appointment = require('../models/Appointment');
const { authenticateJWT, requireRole } = require('../middleware/auth');

// Get all appointments (Doctor-scoped on backend)
router.get('/', authenticateJWT, async (req, res) => {
  try {
    let query = {};
    if (req.user.role === 'doctor') {
      query.doctor_id = req.user.doctor_id;
    }
    const list = await Appointment.find(query).sort({ date: 1, time: 1 });
    res.json(list);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: true, detail: 'Internal server error' });
  }
});

// Book appointment - Admin & Receptionist
router.post('/', authenticateJWT, requireRole('admin', 'receptionist'), async (req, res) => {
  const { patient_id, patient_name, doctor_id, doctor_name, date, time, type, status, notes } = req.body;

  if (!patient_id || !doctor_id || !date || !time) {
    return res.status(422).json({ error: true, detail: 'Missing required fields: patient_id, doctor_id, date, time' });
  }

  try {
    // Conflict slot check: same doctor, date, time, not cancelled
    const conflict = await Appointment.findOne({
      doctor_id: parseInt(doctor_id, 10),
      date,
      time,
      status: { $ne: 'Cancelled' },
    });

    if (conflict) {
      return res.status(409).json({
        error: true,
        detail: `This slot is already booked for ${conflict.patient_name}`,
      });
    }

    const appointment = new Appointment({
      patient_id: String(patient_id),
      patient_name,
      doctor_id: parseInt(doctor_id, 10),
      doctor_name,
      date,
      time,
      type: type || 'Consultation',
      status: status || 'Pending',
      notes: notes || '',
    });

    await appointment.save();
    res.status(201).json(appointment);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: true, detail: 'Internal server error' });
  }
});

// Edit appointment - Admin only
router.put('/:id', authenticateJWT, requireRole('admin'), async (req, res) => {
  const { id } = req.params;
  const updateFields = req.body;

  try {
    const appointment = await Appointment.findById(id);
    if (!appointment) {
      return res.status(404).json({ error: true, detail: 'Appointment not found' });
    }

    const newDoctor = updateFields.doctor_id !== undefined ? parseInt(updateFields.doctor_id, 10) : appointment.doctor_id;
    const newDate = updateFields.date || appointment.date;
    const newTime = updateFields.time || appointment.time;
    const newStatus = updateFields.status || appointment.status;

    // Double booking conflict validation when updating date/time/doctor
    if (newStatus !== 'Cancelled') {
      const conflict = await Appointment.findOne({
        _id: { $ne: id },
        doctor_id: newDoctor,
        date: newDate,
        time: newTime,
        status: { $ne: 'Cancelled' },
      });

      if (conflict) {
        return res.status(409).json({
          error: true,
          detail: `This slot is already booked for ${conflict.patient_name}`,
        });
      }
    }

    // Cast fields to correct types if present
    if (updateFields.doctor_id !== undefined) updateFields.doctor_id = parseInt(updateFields.doctor_id, 10);
    if (updateFields.patient_id !== undefined) updateFields.patient_id = String(updateFields.patient_id);

    // Apply updates
    Object.assign(appointment, updateFields);
    await appointment.save();
    res.json(appointment);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: true, detail: 'Internal server error' });
  }
});

// Cancel appointment - Admin only
router.delete('/:id', authenticateJWT, requireRole('admin'), async (req, res) => {
  const { id } = req.params;

  try {
    const appointment = await Appointment.findById(id);
    if (!appointment) {
      return res.status(404).json({ error: true, detail: 'Appointment not found' });
    }

    appointment.status = 'Cancelled';
    await appointment.save();
    res.json(appointment);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: true, detail: 'Internal server error' });
  }
});

module.exports = router;
